src/Network/WebSockets/Extensions/PermessageDeflate.hs

```diff
@@ -0,0 +1,198 @@
+{-# LANGUAGE OverloadedStrings
+           , LambdaCase
+           , RecordWildCards
+           , BangPatterns
+           , TupleSections
+           #-}
+module Network.WebSockets.Extensions.PermessageDeflate where
+
+import qualified Data.ByteString                  as  BS
+import qualified Data.ByteString.Char8            as BS8
+import qualified Data.ByteString.Lazy.Char8       as BSL8
+import qualified Data.ByteString.Lazy             as BSL
+import qualified Data.ByteString.Lazy.Internal    as   L
+import           Network.WebSockets.Types
+import           Network.WebSockets.Http
+import           Data.Monoid
+import           Data.Either
+import           Control.Applicative              ((<|>), (<$>), (*>), (<*))
+import           Control.Concurrent.MVar
+import           Control.Monad                    (when, liftM2)
+import           Data.Streaming.Zlib
+import qualified Data.Attoparsec.ByteString       as  A hiding (Parser)
+import qualified Data.Attoparsec.ByteString.Char8 as  A
+import qualified Data.CaseInsensitive             as CI
+import           Control.Exception.Safe           (throwIO)
+{-
+   Four extension parameters are defined for "permessage-deflate" to
+   help endpoints manage per-connection resource usage.
+   o  "server_no_context_takeover"
+   o  "client_no_context_takeover"
+   o  "server_max_window_bits"
+   o  "client_max_window_bits"
+-}
+
+data PermessageDeflate = PermessageDeflate
+  { serverNoContextTakeover :: Bool
+  , clientNoContextTakeover :: Bool
+  , serverMaxWindowBits     :: Int
+  , clientMaxWindowBits     :: Int
+  , pdCompressionLevel      :: Int
+  } deriving Show
+
+defaultPermessageDeflate :: PermessageDeflate
+defaultPermessageDeflate = PermessageDeflate False False 15 15 8
+-- defaultPermessageDeflate = PermessageDeflate True True 8 8 9
+
+type UpdatePermessageDeflate = PermessageDeflate -> PermessageDeflate
+
+toClient :: PermessageDeflate -> BS.ByteString
+toClient PermessageDeflate{..} = BS.intercalate "; " $
+     ["permessage-deflate"]
+      ++ ["server_no_context_takeover" | serverNoContextTakeover ]
+      ++ ["client_no_context_takeover" | clientNoContextTakeover ]
+      ++ [BS8.pack $ "server_max_window_bits=" ++ show serverMaxWindowBits | serverMaxWindowBits /= 15 ]
+      ++ [BS8.pack $ "client_max_window_bits=" ++ show clientMaxWindowBits | clientMaxWindowBits /= 15 ]
+
+negotiateDeflate
+  :: Maybe BS8.ByteString
+  -> Maybe PermessageDeflate
+  -> Either BS8.ByteString (Headers, Maybe PermessageDeflate)
```

Comment:
The return type of `negotiateDeflate` is a bit unclear.  It is not obvious that the `BS8.ByteString` contains an error message.  I think there are 3 options:

- Use `Either String (Headers, Maybe PermessageDeflate)`
- Use `Either HandshakeException (Headers, Maybe PermessageDeflate)`
- Use a custom, more verbose type.